functions/go/apply-setters/apply_setters.go

```diff
@@ -0,0 +1,273 @@
+package main
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"sigs.k8s.io/kustomize/kyaml/errors"
+	"sigs.k8s.io/kustomize/kyaml/kio"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+var _ kio.Filter = ApplySetters{}
+
+// ApplySetters applies the setter values to the resource fields which are tagged
+// by the setter reference comments
+type ApplySetters struct {
+	// Setters holds the user provided values for all the setters
+	Setters []Setter `json:"setters,omitempty" yaml:"setters,omitempty"`
+}
+
+type Setter struct {
+	// Name is the name of the setter
+	Name string `json:"name,omitempty" yaml:"name,omitempty"`
+
+	// Value is the input value for setter
+	Value string `json:"value,omitempty" yaml:"value,omitempty"`
+}
+
+const SetterCommentIdentifier = "# kpt-set: "
+
+// Filter implements Set as a yaml.Filter
+func (as ApplySetters) Filter(nodes []*yaml.RNode) ([]*yaml.RNode, error) {
+	for i := range nodes {
+		err := accept(&as, nodes[i])
+		if err != nil {
+			return nil, errors.Wrap(err)
+		}
+	}
+	return nodes, nil
+}
+
+// visitMapping takes input mapping node, and performs following steps
+// checks if the key node of the input mapping node has line comment with SetterCommentIdentifier
+// checks if the value node is of sequence node type
+// if yes to both, resolves the setter value for the setter name in the line comment
+// replaces the existing sequence node with the new values provided by user
+
+/*
+e.g. for input of Mapping node
+environments: # kpt-set: ${env}
+- dev
+- stage
+
+and for input ApplySetters [name: env, value: "[stage, prod]"]
+
+The yaml node is transformed to
+environments: # kpt-set: ${env}
+- stage
+- prod
+*/
+func (as *ApplySetters) visitMapping(object *yaml.RNode) error {
+	return object.VisitFields(func(node *yaml.MapNode) error {
+		if node.IsNilOrEmpty() {
+			return nil
+		}
+		// the aim of this method is to apply-setter for sequence nodes
+		if node.Value.YNode().Kind != yaml.SequenceNode {
+			// return if it is not a sequence node
+			return nil
+		}
+
+		setterPattern := extractSetterPattern(node.Key)
+		if setterPattern == "" {
+			// the node is not tagged with setter pattern
+			return nil
+		}
+
+		if !shouldSet(setterPattern, as.Setters) {
+			// this means there is no intent from user to modify this setter tagged resources
+			return nil
+		}
+
+		// since this setter pattern is found on sequence node, make sure that it is
+		// not interpolation of setters, it should be simple setter e.g. ${environments}
+		if !validArraySetterPattern(setterPattern) {
+			return errors.Errorf("setter interpolation is not allowed for arrays: %q", setterPattern)
```

Comment:
Ideally, this should never happen because Search and Replace wouldn't allow interpolation. This case might arise when package publishers add incorrect setter comments manually. Changed the error message to be generic to avoid interpolation term. Package consumers need not know about interpolation.